File: c/grains/grains.c
#include "grains.h"
#include <math.h>
#include <stdint.h>

uint64_t square(uint8_t index) {
  uint64_t count = pow(2, index - 1);
  return count;
}

uint64_t total() {
  uint64_t count = 0;
  for (int i = 1; i <= 64; i++) {
    count += square(i);
  }
  return count;
}

File: zig/grains/grains.zig
const pow = @import("std").math.pow;
const max = @import("std").math.maxInt;
pub const ChessboardError = error{IndexOutOfBounds};

pub fn square(index: usize) ChessboardError!u64 {
    if (index > 64 or index < 1) return ChessboardError.IndexOutOfBounds;
    return @as(u64, 1) << @intCast(index - 1);
}

pub fn total() u64 {
    return max(u64);
}
